### PR TITLE
Add support for approving InstallPlans with multiple CSVs

### PR DESCRIPTION
**Pull request by mprahl**
# First Commit

The operatorpolicy.policy.open-cluster-management.io/managed label is set to an empty value on the Subscription. The
operatorpolicy.policy.open-cluster-management.io/managed annotation is set to <policy name>.<policy namespace>.

The label doesn't set a value because the value could be too long.

This will allow querying for subscriptions managed by operator policies.

# Second Commit

When an InstallPlan contains multiple CSVs, it will be approved if all CSVs can be approved by the current operator policies.

Rather than query the operator policies directly, the subscriptions in the namespace that are managed by an operator policy are queried and then their associated operator policies are queried to avoid having to resolve templates.

Relates:
https://issues.redhat.com/browse/ACM-11981

### CODE REVIEW

File: controllers/operatorpolicy_controller.go
Comment (JustinKuli) [inline controllers/operatorpolicy_controller.go:1512]: I understand prioritizing the Installing plan, but why also Complete plans? It feels like there can be multiple old InstallPlans that have that one, if the operator is done installing?

Comment (mprahl) [reply controllers/operatorpolicy_controller.go:1512]: @JustinKuli if an InstallPlan is complete for `currentCSV` (not `installedCSV`) then that means this InstallPlan successfully installed the latest CSV in the channel for this subscription. Otherwise, due to OLM race conditions, it's possible that there are multiple InstallPlans for `currentCSV` where one is approved and completed and one is awaiting approval.

File: controllers/operatorpolicy_controller.go
Comment (JustinKuli) [inline controllers/operatorpolicy_controller.go:1795]: As a weird note, previously `allowedVersions` was the variable name.

In this review, it is difficult to tell what is new-new code and what is small changes to existing pieces. At least in GH, it's not very good at aligning the old and new... This is obviously not a major issue, but if there were some additional "anchor points" where there weren't changes (these six lines, for example), it might help `git blame` in the future.

File: controllers/operatorpolicy_controller.go
Comment (JustinKuli) [inline controllers/operatorpolicy_controller.go:1802]: Is there a guarantee that `sub.Status.CurrentCSV` is in `csvNames`?

Comment (JustinKuli) [reply controllers/operatorpolicy_controller.go:1802]: It feels like there could be items in `csvNames` that "should" be approved by a policy with no versions specified - I thought we were considering some sort of name matching in this case. Is that somewhere else? Or is it not necessary?

File: controllers/operatorpolicy_controller.go
Comment (mprahl) [inline controllers/operatorpolicy_controller.go:1591]: @JustinKuli I read the OLM code and it seems that the only successful final states are created (even if it updated) and present. Do you think it's best to do this filtering or should we just mark the InstallPlan as failed even if this operator was successful?

I don't really know how I can get test coverage on this because `clusterServiceVersionNames` order in the InstallPlan are not deterministic so even if I have an InstallPlan with a failing operator, that one may have been processed first and not move on to the successful operator. I suppose I could modify the InstallPlan status to order the expected successful one but that seems quite hacky.

Comment (JustinKuli) [reply controllers/operatorpolicy_controller.go:1599]: I didn't fully understand this part, so if you think it can work without it, that would be fine? I don't like the idea of modifying the InstallPlan status

File: controllers/operatorpolicy_controller.go
Comment (mprahl) [reply controllers/operatorpolicy_controller.go:1802]: Here are the breadcrumbs:
https://github.com/operator-framework/operator-lifecycle-manager/blob/889ae50a66fe36a46a3c756a0ad7b481b73672f1/pkg/controller/operators/catalog/operator.go#L1291
https://github.com/operator-framework/operator-lifecycle-manager/blob/889ae50a66fe36a46a3c756a0ad7b481b73672f1/pkg/controller/registry/resolver/step_resolver.go#L123-L127
https://github.com/operator-framework/operator-lifecycle-manager/blob/889ae50a66fe36a46a3c756a0ad7b481b73672f1/pkg/controller/registry/resolver/step_resolver.go#L132-L136
https://github.com/operator-framework/operator-lifecycle-manager/blob/889ae50a66fe36a46a3c756a0ad7b481b73672f1/pkg/controller/registry/resolver/step_resolver.go#L197-L204
https://github.com/operator-framework/operator-lifecycle-manager/blob/889ae50a66fe36a46a3c756a0ad7b481b73672f1/pkg/controller/operators/catalog/operator.go#L1428-L1433

File: controllers/operatorpolicy_controller.go
Comment (mprahl) [reply controllers/operatorpolicy_controller.go:1599]: @JustinKuli to clarify, modifying the InstallPlan status would be for test coverage.

So the question is, should an InstallPlan be marked as failed in the OperatorPolicy if the overall InstallPlan failed but the part for the operator being managed by OperatorPolicy succeeded?

File: controllers/operatorpolicy_controller.go
Comment (mprahl) [reply controllers/operatorpolicy_controller.go:1795]: The side by side diff is easier for this.

File: controllers/operatorpolicy_controller.go
Comment (mprahl) [reply controllers/operatorpolicy_controller.go:1802]: @JustinKuli I'll revise my statement of "latest in the channel" but rather the latest that replaces the current version.

**Review by JustinKuli — APPROVED** [review]
/hold

**Pull request merged**
